### PR TITLE
Story #1934089: SDK update to version 16.0.400.1

### PR DESCRIPTION
**Pull request by tatizapirtan**
https://center.almoctane.com/ui/entity-navigation?p=1001/1002&entityType=work_item&id=1934089
Story is about updating the sdk version from 12.60.60 to 16.0.400.1

Biggest issue was with the Authentication class from sdk. 
Before it looked like this:
image
Now it looks like this:
image
Fortunately getClientHeader() wasn't used anywhere, just getAuthenticationString() was used in 1 place and even more fortunately there was a subclass of Authentication, called JSONAuthentication which has this method. And that's why you'll see a lot of changes from Authentication to JSONAuthentication
image

Plugin has 2 methods for authentication, one with user credentials and one via browser. For the first one we have the class UserAuthentication which will inherit from SimpleUserAuthentication, just like before but with a change in method names (username becomes authenticationId and password becomes authenticationSecret) and for the second one we have GrantTokenAuthentication which acts more like a marker class and which i changed to extend SimpleClientAuthentication since this was the most appropriate class to use.

Another issue was with method authenticate() from OctaneHttpClient which is overrode in IdePluginsOctaneHttpClient.java. Before it received an authentication parameter but now no parameters are passed. In order to not disturb the existing logic i added a setter for this authentication parameter which is called before calling the autenticate() method

### CODE REVIEW

File: src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java
Comment (vladbenchea) [inline src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java:137]: I'm a bit confused - now that you don't assign any value to lastUsedAuthentication when first entering this method, doesn't this check throw  a NPE or other error?

File: pom.xml
Comment (vladbenchea) [inline pom.xml:160]: why is this needed? I haven't seen any usage in this PR

Comment (tatizapirtan) [reply pom.xml]: It was needed because of this error:
image

The google-http-client library was calling this method from guava. But guava is a library which many other libraries are using therefore it was imported from multiple places with different versions. The version used in our projects was 16 which did not implement that method. That method existed in guava only from version 20. Therefore first i thought to just add the guava library itself with the latest version. However this did not work in eclipse plugin. So the final working solution was to increase the version of guice library to 4.2.0 which brings guava 23.6

File: src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java
Comment (tatizapirtan) [reply src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java:137]: No because i use a setter now to initialize lastUsedAuthentication which i usually call before calling this method

Comment (vladbenchea) [reply src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java:137]: usually? are there any cases where a NPE can be thrown?

Comment (tatizapirtan) [reply src/main/java/com/hpe/adm/octane/ideplugins/services/connection/IdePluginsOctaneHttpClient.java:137]: The authenticate method is called inside this IdePluginsOctaneHttpClient class and from outside. When it's called from outside i use the setter and when it's called from inside there is a null check before. So no, there shouldn't be any cases where NPE can be thrown

**Review by vladbenchea — APPROVED** [review]
Please wait for Dan's approval as well

**Review by pikachugb — APPROVED** [review]

**Pull request merged**
